src/Formats/IonReader.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_ION
+
+#include <IO/ReadBuffer.h>
+
+#include <ionc/ion.h>
+
+namespace DB
+{
+
+class ReadBuffer;
+
+using IonDecimalPtr = std::shared_ptr<ION_DECIMAL>;
+using IonIntPtr = std::shared_ptr<ION_INT>;
+using IonTimestampPtr = std::shared_ptr<ION_TIMESTAMP>;
+
+using Symbol = struct
+{
+    std::string value;
+};
+
+struct IonNull
+{
+    ION_TYPE tpe;
+};
+
+const int MAX_DECIMAL_DIGITS = 10000;
+
+class IonReader
+{
+private:
+    hREADER reader;
+    ION_TYPE current_type;
+    decContext current_decimal_context;
+
+    ReadBuffer & in;
+
+public:
+    IonReader(ReadBuffer & in_);
+    ~IonReader();
+
+    template <typename T>
+    T read() {
+        T val;
+        auto err = read(val);
+        if (err != IERR_OK)
+            printf("ERROR reading value: %d\n", err);
```

Comment:
We should throw an exception in case of error. Something like
```
throw Exception(ErrorCodes::INCORRECT_DATA, "Cannot read data in ION format {}", <some_text_message_from_err>)
```